l1-contracts/contracts/chain-registrar/ChainRegistrar.sol

```diff
@@ -0,0 +1,196 @@
+
+    /// @dev Thrown when the bridge for a chain is not registered.
+    error BridgeIsNotRegistered();
+
+    /// @notice Emitted when a new chain registration proposal is made.
+    /// @param chainId Unique ID of the proposed chain.
+    /// @param author Address of the proposer.
+    event NewChainRegistrationProposal(uint256 indexed chainId, address author);
+
+    /// @notice Emitted when the L2 deployer address is changed.
+    /// @param newDeployer Address of the new L2 deployer.
+    event L2DeployerChanged(address newDeployer);
+
+    /// @dev Struct for holding the base token configuration of a chain.
+    struct BaseToken {
+        /// @notice Gas price multiplier numerator, used to compare the base token price to ether for L1->L2 transactions.
+        uint128 gasPriceMultiplierNominator;
+        /// @notice Gas price multiplier denominator, used to compare the base token price to ether for L1->L2 transactions.
+        uint128 gasPriceMultiplierDenominator;
+        /// @notice Address of the base token used for gas fees.
+        address tokenAddress;
+        /// @notice Address responsible for setting the token multiplier.
+        address tokenMultiplierSetter;
+    }
+
+    /// @dev Struct for holding the configuration of a proposed chain.
+    // solhint-disable-next-line gas-struct-packing
+    struct ChainConfig {
+        /// @notice Unique chain ID.
+        uint256 chainId;
+        /// @notice Base token configuration for the chain.
+        BaseToken baseToken;
+        /// @notice Operator responsible for making commit transactions.
+        address blobOperator;
+        /// @notice Operator responsible for making prove and execute transactions.
+        address operator;
+        /// @notice Governor of the chain; will receive ownership of the ChainAdmin contract.
+        address governor;
+        /// @notice Mode for charging users for pubdata.
+        PubdataPricingMode pubdataPricingMode;
+    }
+
+    /// @dev Struct for holding the configuration of a fully deployed chain.
+    // solhint-disable-next-line gas-struct-packing
+    struct RegisteredChainConfig {
+        /// @notice Address of the pending admin for the chain.
+        address pendingChainAdmin;
+        /// @notice Address of the current admin for the chain.
+        address chainAdmin;
+        /// @notice Address of the main contract (diamond proxy) for the deployed chain.
+        address diamondProxy;
+        /// @notice Address of the L2 bridge inside the deployed chain.
+        address l2BridgeAddress;
+    }
+
+    /// @notice Initializes the contract with the given parameters.
+    /// @dev Can only be called once, during contract deployment.
+    /// @param _bridgehub Address of the ZKsync Bridgehub.
+    /// @param _l2Deployer Address of the L2 deployer.
+    /// @param _owner Address of the contract owner.
+    function initialize(address _bridgehub, address _l2Deployer, address _owner) public {
```

Comment:
Important: anyone can reinitialize the contract. To avoid it you can use `Initializable.sol` from OZ lib. Also don't forget to set `_disableInitializers` in constructor.
```suggestion
    function initialize(address _bridgehub, address _l2Deployer, address _owner) external Initializer {
```